sphinx/ext/autodoc/__init__.py

```diff
@@ -1341,20 +1354,21 @@ def get_user_defined_function_or_method(obj: Any, attr: str) -> Any:
         # the signature from, so just pass the object itself to our hook.
         self.env.app.emit('autodoc-before-process-signature', self.object, False)
         try:
-            return inspect.signature(self.object, bound_method=False)
+            sig = inspect.signature(self.object, bound_method=False)
+            return None, None, sig
         except ValueError:
             pass
 
         # Still no signature: happens e.g. for old-style classes
         # with __init__ in C and no `__text_signature__`.
-        return None
+        return None, None, None
 
     def format_args(self, **kwargs: Any) -> str:
         if self.env.config.autodoc_typehints in ('none', 'description'):
             kwargs.setdefault('show_annotation', False)
 
         try:
-            sig = self._get_signature()
+            _, _, sig = self._get_signature()
```

Comment:
What I was proposing earlier was keeping the hack of attaching attributes to self, but moving it out of `_get_signature` and into here.

The problem with calling `_get_signature` twice is that the user's hooks get called twice.

Reply:
oops. I have to change it again.